### ISSUE TITLE
subspan modulo spuriously fills across missing cells adjacent to "ghost cell"

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#1638]
**Reported by andrew wittenberg on 10 Feb 2016 22:31 UTC**
For subspan modulo axes, note how @fln spuriously fills across the NetCDF-specified missing values, but not across the Ferret-inserted "ghost cell".  Ferret needs some logic for what to do when a run of one or more missing cells are adjacent to the "ghost cell."  

```
 	NOAA/PMEL TMAP
 	FERRET v6.97 (PyFerret 1.3)
 	Linux 2.6.32-573.1.1.el6.x86_64 - 01/21/16
 	10-Feb-16 17:14     

yes? def ax/x=0:180:10/unit=degree_e xax
yes? let a = x[save/clob/file=a.nc a
 LISTing to file a.nc
yes? let b = IF a GE 50 THEN a
yes? save/clob/file=b.nc b
 LISTing to file b.nc
yes? can var/all; can ax xax
yes? use b
yes? sho dat
     currently SET data sets:
    1> ./b.nc  (default)
 name     title                             I         J         K         L         M         N
 B        IF A GE 50 THEN A                1:19      ...       ...       ...       ...       ...
 
yes? list/x=-180:720 b
             VARIABLE : IF A GE 50 THEN A
             FILENAME : b.nc
             SUBSET   : 43 points (LONGITUDE)
 180E    / -1:  180.0
 90W     /  0:   ....
 0E      /  1:   ....
 10E     /  2:   ....
 20E     /  3:   ....
 30E     /  4:   ....
 40E     /  5:   ....
 50E     /  6:   50.0
 60E     /  7:   60.0
 70E     /  8:   70.0
 80E     /  9:   80.0
 90E     / 10:   90.0
 100E    / 11:  100.0
 110E    / 12:  110.0
 120E    / 13:  120.0
 130E    / 14:  130.0
 140E    / 15:  140.0
 150E    / 16:  150.0
 160E    / 17:  160.0
 170E    / 18:  170.0
 180E    / 19:  180.0
 90W     / 20:   ....
 0E(360) / 21:   ....
 10E     / 22:   ....
 20E     / 23:   ....
 30E     / 24:   ....
 40E     / 25:   ....
 50E     / 26:   50.0
 60E     / 27:   60.0
 70E     / 28:   70.0
 80E     / 29:   80.0
 90E     / 30:   90.0
 100E    / 31:  100.0
 110E    / 32:  110.0
 120E    / 33:  120.0
 130E    / 34:  130.0
 140E    / 35:  140.0
 150E    / 36:  150.0
 160E    / 37:  160.0
 170E    / 38:  170.0
 180E    / 39:  180.0
 90W     / 40:   ....
 0E(720) / 41:   ....
yes? list/x=-180:720 b[x=@fln](gx=xax]
yes?)
             VARIABLE : IF A GE 50 THEN A
                        linear-filled by 1 pts on X
             FILENAME : b.nc
             SUBSET   : 43 points (LONGITUDE)
 180E    / -1:  180.0
 90W     /  0:   ....
 0E      /  1:   78.3
 10E     /  2:   72.6
 20E     /  3:   67.0
 30E     /  4:   61.3
 40E     /  5:   55.7
 50E     /  6:   50.0
 60E     /  7:   60.0
 70E     /  8:   70.0
 80E     /  9:   80.0
 90E     / 10:   90.0
 100E    / 11:  100.0
 110E    / 12:  110.0
 120E    / 13:  120.0
 130E    / 14:  130.0
 140E    / 15:  140.0
 150E    / 16:  150.0
 160E    / 17:  160.0
 170E    / 18:  170.0
 180E    / 19:  180.0
 90W     / 20:   ....
 0E(360) / 21:   ....
 10E     / 22:   ....
 20E     / 23:   ....
 30E     / 24:   ....
 40E     / 25:   ....
 50E     / 26:   50.0
 60E     / 27:   60.0
 70E     / 28:   70.0
 80E     / 29:   80.0
 90E     / 30:   90.0
 100E    / 31:  100.0
 110E    / 32:  110.0
 120E    / 33:  120.0
 130E    / 34:  130.0
 140E    / 35:  140.0
 150E    / 36:  150.0
 160E    / 37:  160.0
 170E    / 38:  170.0
 180E    / 39:  180.0
 90W     / 40:   ....
 0E(720) / 41:   ....
```

Interestingly, this only happens for the first modulo branch, not the second.  But @fnr does it for both branches:

```
yes? list/x=-180:720 b[x=@fnr]
             VARIABLE : IF A GE 50 THEN A
                        nearest-filled by 1 pts on X
             FILENAME : b.nc
             SUBSET   : 43 points (LONGITUDE)
 180E    / -1:  180.0
 90W     /  0:   ....
 0E      /  1:  180.0
 10E     /  2:  180.0
 20E     /  3:   50.0
 30E     /  4:   50.0
 40E     /  5:   50.0
 50E     /  6:   50.0
 60E     /  7:   60.0
 70E     /  8:   70.0
 80E     /  9:   80.0
 90E     / 10:   90.0
 100E    / 11:  100.0
 110E    / 12:  110.0
 120E    / 13:  120.0
 130E    / 14:  130.0
 140E    / 15:  140.0
 150E    / 16:  150.0
 160E    / 17:  160.0
 170E    / 18:  170.0
 180E    / 19:  180.0
 90W     / 20:   ....
 0E(360) / 21:   50.0
 10E     / 22:   50.0
 20E     / 23:   50.0
 30E     / 24:   50.0
 40E     / 25:   50.0
 50E     / 26:   50.0
 60E     / 27:   60.0
 70E     / 28:   70.0
 80E     / 29:   80.0
 90E     / 30:   90.0
 100E    / 31:  100.0
 110E    / 32:  110.0
 120E    / 33:  120.0
 130E    / 34:  130.0
 140E    / 35:  140.0
 150E    / 36:  150.0
 160E    / 37:  160.0
 170E    / 38:  170.0
 180E    / 39:  180.0
 90W     / 40:   ....
 0E(720) / 41:   ....

```

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/2366

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 11 Feb 2016 18:13 UTC**
Just jotting down a 'hunch' here, as I am busy on other tasks:

The "void' cell that the subspan modulo inserts is just a grid point filled with a missing value flag -- nothing more.  Our goal is to get all possible transformations -- fillers, smoothers, derivatives, etc. -- to recognize this void as special -- NO NOT CROSS THIS LINE.  But there may be, nothing special in what these transformations see in the data, as the modulo-ing of the data may have occurred many layers down in a pyramid of definitions.  Furthermore, we need to consider user-written external functions.  We cannot ask those functions to be on the lookout for phantom void points.

The right solution to this problem, I suspect, is in INTERP_STACK.  We need Ferret's evaluation stack to be made aware that it has been asked to act upon an X (or T) axis span that contains a void, and to split that operation into two separate and disjoint invocations, and then rejoin them when received back -- much like IS_GATHER does for splitting up overly large calculations

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 11 Feb 2016 18:33 UTC**
Exactly.  The transformations do not know that they are dealing with a special case.

One thing to check first, is the locations of the grid cell boundaries for the ghost cell. Are they defined so the cells at the edges of the actual data are of the same size as the rest of the data, or are the boundaries halfway between the ghost cell point and the last coordinate point of the data. If the void point is causing the edge of the data to be seen as in a huge cell, this might account for a lot of the bad behavior.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 11 Feb 2016 19:58 UTC**

''
One thing to check first, is the locations of the grid cell boundaries for the ghost cell. Are they defined so the cells at the edges of the actual data are of the same size as the rest of the data, or are the boundaries halfway between the ghost cell point and the last coordinate point of the data. If the void point is causing the edge of the data to be seen as in a huge cell, this might account for a lot of the bad behavior.
''

This is fine (of course).

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 11 Feb 2016 23:24 UTC**
Quick look in the code of INTERP_STACK makes this look like a simple fix.  What we see is:

```
* ... strip off a transformation ?
   [to strip off transformations](code)
...
* ... strip off "modulo" replications
   [to strip off traditional (globe-encircling axis) modulos](code)
...
* ... strip off creation of virtual "void" point on subspan modulo axes
    IF (TM_ITS_SUBSPAN_MODULO(CGRID_AXIS(idim, cx))) THEN
  ...
```

The third of these -- "creation of virtual "void" point on subspan modulo axes" -- divides the full region into two separate requests, and then re-assembles the results on the target axis with the void point filled in between the two pieces.

Since the order of execution is the reverse of the order in which operations are stripped off, this code builds the subspan modulo representation of the data and then performs transformations (e.g. smoothers) upon it.  This allows the undesirable behavior of the smoother smearing the data across the void.

To fix the problem I believe we need only ensure that we strip off the subspan modulo operation before any other operations are stripped off.  That way each side of the void is computed with complete independence.  There is no possibility of information bleeding across the void.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 5 May 2016 17:47 UTC**
This is fixed, along with #2364.
